frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+          notificationType,
+          notificationType === MESSAGE_NOTIFICATIONS.ADD_MEMBER ? { username } : {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        state.messages.push(newMessage)
+
+        Vue.set(state.users, username, { joinedDate: meta.createdDate })
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvent({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/rename': {
+      validate: objectOf({
+        name: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        Vue.set(state.attributes, 'name', data.name)
+        const notificationData = createNotificationData(MESSAGE_NOTIFICATIONS.UPDATE_NAME, {})
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        state.messages.push(newMessage)
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvent({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/changeDescription': {
+      validate: objectOf({
+        description: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        Vue.set(state.attributes, 'description', data.description)
+        const notificationData = createNotificationData(
+          MESSAGE_NOTIFICATIONS.UPDATE_DESCRIPTION, {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        state.messages.push(newMessage)
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvent({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/leave': {
+      validate: objectOf({
+        username: optional(string), // coming from the gi.contracts/group/leaveChatRoom
+        member: string // username to be removed
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { member } = data
+        const isKicked = data.username && member !== data.username
+        if (!state.users[member]) {
+          throw new Error(`Can not leave the chatroom which ${member} are not part of`)
+        }
+        Vue.delete(state.users, member)
+
+        const notificationType = !isKicked ? MESSAGE_NOTIFICATIONS.LEAVE_MEMBER : MESSAGE_NOTIFICATIONS.KICK_MEMBER
+        const notificationData = createNotificationData(notificationType, isKicked ? { username: member } : {})
+        const newMessage = createMessage({
+          meta: isKicked ? meta : { ...meta, username: member },
+          hash,
+          data: notificationData,
+          state
+        })
+        state.messages.push(newMessage)
+      },
+      sideEffect ({ data, hash, contractID }, { state }) {
+        if (sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+          return
+        }
```

Comment:
This check is incorrectly positioned.

What you want to do is check to see if the user that is leaving this chatroom is also the user that is joining this chatroom right now.

However, the way this is written here, if we are joining any chatroom at all (including a different chatroom), and we get a `'leave'` action for a different user, it won't properly run the code below as it should.

You need to move this check inside of the `if` below.

Reply:
Nice catch.